src-tauri/crates/spawelo/src/lib.rs

```diff
@@ -78,99 +42,96 @@ pub fn ml_elo(
         ML_ITERATIONS
     );
 
-    let start = Instant::now();
-    let mut scores = vec![];
+    let start: Instant = Instant::now();
     for i in 0..ML_ITERATIONS {
-        let elo_sum: i64 = elo.values().sum();
-        let noisy_history = add_noise(history);
-
+        let elo_sum: f64 = elo.values().sum();
         print_debug(i, history, &elo, elo_sum);
 
-        let mut rng = rand::thread_rng();
-        let rand_player = elo.keys().choose(&mut rng).unwrap();
-        let score_adj = Uniform::new_inclusive(-30, 30).sample(&mut rng);
-        let mut new_elo = elo.clone();
-        *new_elo.get_mut(rand_player).unwrap() += score_adj;
-
-        let old_score =
-            probability_of_perfect_prediction(&noisy_history, &elo) * regularization_penalty(&elo);
-        let new_score = probability_of_perfect_prediction(&noisy_history, &new_elo)
-            * regularization_penalty(&new_elo);
-        if new_score > old_score {
-            elo = new_elo;
-        }
-        scores.push(new_score);
-
-        if should_stop_iteration(i, history, &elo) {
-            break;
+        let derivative: HashMap<PlayerId, f64> = backpropagation(history, &elo);
+        for (player, diff) in derivative {
+            *elo.entry(player).or_default() += diff * LEARNING_RATE;
         }
     }
+
     log_elo(&elo);
     log_probabilities(&elo, history);
     info!("ELO calculations took {:?}", start.elapsed());
-    elo.into_iter().map(|i| (i.0.clone(), i.1)).collect()
+
+    elo
 }
 
-fn log_elo(elo: &HashMap<&PlayerId, i64>) {
+fn log_elo(elo: &HashMap<PlayerId, f64>) {
     let mut elo: Vec<_> = elo.into_iter().collect();
-    elo.sort_by_key(|p| p.1);
+    elo.sort_by_key(|p| *p.1 as i64 * 1000);
```

Comment:
Isn't it better to multiply before casting to int?

Reply:
Multiplying after cast is rather pointless.

Reply:
Fixed